lib/rex/post/meterpreter/packet_dispatcher.rb

```diff
@@ -729,9 +732,7 @@ def shutdown_passive_dispatcher
       resource_uri = "/" + self.conn_id.to_s.gsub(/(^\/|\/$)/, '') + "/"
       self.passive_service.remove_resource(resource_uri) if self.passive_service
 
-      if self.passive_service.resources.empty?
-        Rex::ServiceManager.stop_service(self.passive_service)
-      end
+      Rex::ServiceManager.stop_service(self.passive_service)
```

Comment:
Just confirming, should this be a `deref` call rather than an explicit call to `stop_manager` ? 👀

Reply:
Two http sessions opened:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > sessions

Active sessions
===============

  Id  Name  Type                 Information               Connection
  --  ----  ----                 -----------               ----------
  1         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()
  2         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()

```

services:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > _servicemanager 
Services
========

 Id  Name                                         References
 --  ----                                         ----------
 0   HTTP Server                                  4
 1   Rex::Proto::Http::Server8080-192.168.123.1-  4
```

Killing the first session:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > sessions -k 1
[*] Killing the following session(s): 1
[*] Killing session 1

[*]  - Meterpreter session 1 closed.
```

Confirming first session killed:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > sessions

Active sessions
===============

  Id  Name  Type                 Information               Connection
  --  ----  ----                 -----------               ----------
  2         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()
```

But now the service manager is empty:

```
msf6 payload(osx/x64/meterpreter_reverse_http) > _servicemanager 
[*] No framework services are currently running.
```

the listener is still open:
```
lsof -i :8080
COMMAND   PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
ruby    23249 a   14u  IPv4 0xea596340b33920c7      0t0  TCP 192.168.123.1:http-alt (LISTEN)
```

And it's no longer possible to create a new handler:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > to_handler
[*] Payload Handler Started as Job 1
msf6 payload(osx/x64/meterpreter_reverse_http) > 
[-] Handler failed to bind to 192.168.123.1:8080
[*] Started HTTP reverse handler on http://0.0.0.0:8080/base_url
```

Reply:
Interestingly even though the service manager claims to be empty with no services, you can still open new sessions:

```
msf6 payload(osx/x64/meterpreter_reverse_http) > _servicemanager 
[*] No framework services are currently running.
```

Opening a new payload continues to work:

```
msf6 payload(osx/x64/meterpreter_reverse_http) > [*] Meterpreter session 3 opened (192.168.123.1:8080 -> 127.0.0.1 ) at 2022-03-09 11:39:19 +0000
```

Reply:
Right, this turned out to be a weirdness in the service manager; I'd initially kept the existing convention of services sayping "I'm done" by letting the `ServiceManager` know; but that just had the effect that the ServiceManager no longer kept track of it; but the service was indeed still running. Trying to start a new one would then fail because the ServiceManager no longer knew that it was running; although the deref would at least still clean it up correctly. I've now changed it to just use deref everywhere, and it seems to be behaving a lot more consistently.